store/store.go

```diff
@@ -1,16 +1,57 @@
 package store
 
 import (
+	"bytes"
 	"encoding/json"
 
 	"github.com/ipfs/go-cid"
 	peer "github.com/libp2p/go-libp2p-core/peer"
+	varint "github.com/multiformats/go-varint"
 )
 
 // IndexEntry describes the information to be stored for each CID in the indexer.
 type IndexEntry struct {
-	ProvID  peer.ID // ID of the provider of the CID.
-	PieceID cid.Cid // PieceID of the CID where the CID is stored in the provider (may be nil)
+	// PrividerID is the peer ID of the provider of the CID
+	ProviderID peer.ID
+	// Metadata is serialized data that provides information about retrieving
+	// data, for the indexed CID, from the identified provider.
+	Metadata []byte
+}
+
+func MakeIndexEntry(providerID peer.ID, protocol uint64, data []byte) IndexEntry {
+	return IndexEntry{
+		ProviderID: providerID,
+		Metadata:   encodeMetadata(protocol, data),
+	}
+}
+
+// PutData writes the protocol ID and the encoded data to IndexEntry.Metadata
+func (ie *IndexEntry) PutData(protocol uint64, data []byte) {
+	ie.Metadata = encodeMetadata(protocol, data)
+}
+
+// GetData returns the protocol ID and the encoded data from the IndexEntry.Metadata
+func (ie *IndexEntry) GetData() (uint64, []byte, error) {
+	protocol, len, err := varint.FromUvarint(ie.Metadata)
```

Comment:
Quick question to see if I understood this right. The protocolID determines how the data has been encoded, right? How do you envision the flow of integrating new protocols? Who determines what protocolID to use, the provider or the underlying transport protocol used to attest data to the index?

Reply:
I think the provider will choose this, but I do not think anyone has decided exactly this would work.  I am thinking that if a provider has certain information that helps them look up data for a CID, (deal id, region, storage data, special magic number), then they will register a protocol number and some specification of what data accompanies a retrieve CID request.